src/common/yaml.hpp

```diff
@@ -0,0 +1,165 @@
+using ExpectedInt64 = mender::common::expected::ExpectedInt64;
+using ExpectedDouble = mender::common::expected::ExpectedDouble;
+using ExpectedBool = mender::common::expected::ExpectedBool;
+using ExpectedSize = mender::common::expected::ExpectedSize;
+
+class Yaml {
+public:
+	using ExpectedYaml = expected::expected<Yaml, error::Error>;
+	using ChildrenMap = map<string, Yaml>;
+	using ExpectedChildrenMap = expected::expected<ChildrenMap, error::Error>;
+
+	Yaml() = default;
+
+	Yaml(const Yaml &) = default;
+	Yaml &operator=(Yaml &) = default;
+	Yaml &operator=(const Yaml &) = default;
+
+	Yaml(Yaml &&) = default;
+	Yaml &operator=(Yaml &&) = default;
+
+	string Dump(const int indent = 2) const;
+
+	ExpectedYaml Get(const char *child_key) const;
+	ExpectedYaml operator[](const char *child_key) const {
+		return this->Get(child_key);
+	}
+	ExpectedYaml Get(const string &child_key) const {
+		return this->Get(child_key.data());
+	}
+	ExpectedYaml operator[](const string &child_key) const {
+		return this->Get(child_key.data());
+	}
+	ExpectedYaml Get(const size_t idx) const;
+	ExpectedYaml operator[](const size_t idx) const {
+		return this->Get(idx);
+	}
+
+	ExpectedChildrenMap GetChildren() const;
+
+	bool IsObject() const;
+	bool IsArray() const;
+	bool IsString() const;
+	bool IsInt() const;
+	bool IsNumber() const;
+	bool IsDouble() const;
+	bool IsBool() const;
+	bool IsNull() const;
+
+	template <typename T>
+	expected::expected<T, error::Error> Get() const;
+
+	ExpectedSize GetArraySize() const;
+
+	string GetType() const;
+	friend std::ostream &operator<<(std::ostream &os, const Yaml &y) {
+		os << y.n_yaml;
+		return os;
+	}
+
+	friend ExpectedYaml LoadFromFile(string file_path);
+	friend ExpectedYaml Load(string yaml_str);
+	friend ExpectedYaml Load(istream &str);
+	friend ExpectedYaml Load(io::Reader &reader);
+
+public:
+#ifdef MENDER_USE_YAML_CPP
+	YAML::Node n_yaml;
+	Yaml(YAML::Node n_yaml) :
+		n_yaml(n_yaml) {};
+#endif
+};
+
+using ExpectedYaml = expected::expected<Yaml, error::Error>;
+using ChildrenMap = map<string, Yaml>;
+using ExpectedChildrenMap = expected::expected<ChildrenMap, error::Error>;
+
+ExpectedYaml LoadFromFile(string file_path);
+ExpectedYaml Load(string yaml_str);
+ExpectedYaml Load(istream &str);
+ExpectedYaml Load(io::Reader &reader);
+
+using ExpectedStringVector = expected::ExpectedStringVector;
+using KeyValueMap = unordered_map<string, string>;
+using ExpectedKeyValueMap = expected::expected<KeyValueMap, error::Error>;
+
+ExpectedStringVector ToStringVector(const yaml::Yaml &j);
+ExpectedKeyValueMap ToKeyValueMap(const yaml::Yaml &j);
+ExpectedString ToString(const yaml::Yaml &j);
+ExpectedInt64 ToInt(const yaml::Yaml &j);
+ExpectedBool ToBool(const yaml::Yaml &j);
+
+enum class MissingOk {
+	No,
+	Yes,
+};
+
+template <typename T>
+expected::expected<T, error::Error> Get(
+	const yaml::Yaml &yaml, const string &key, MissingOk missing_ok);
```

Comment:
Why not provide `missing_ok` in the member function and make this a simple forwarder? Seems unnecessary that one would support it and not the other.

Reply:
Yes, struck me as well. But it is our existing interface. I was unsure to what degree I was allowed to be flexible here, so I stuck with it.

All for doing what you say, if it's not a problem

Reply:
Right. It's not a big deal, we can also relax this if we ever get to a point where we get annoyed by it! 😄